### PR TITLE
Fix incorrect use of the function gp_execution_segment

### PR DESCRIPTION
**Pull request by tsdkv**
When we use the function `gp_execution_segment` as joinrel, the scheduler puts its execution on one segment and replicates the result to the others. This does not lead to the expected results.

### CODE REVIEW

File: src/backend/catalog/gp_toolkit.sql [hunk 1]
```diff
@@ -498,14 +498,9 @@ BEGIN
     SELECT gp_toolkit.__gp_param_local_setting($1) INTO paramdummy;
 
     FOR paramsettings IN
-        SELECT gs.*
-        FROM gp_toolkit.__gp_localid, gp_toolkit.__gp_param_local_setting($1) AS gs
-
+        select gp_execution_segment(), $1, current_setting($1)::text from gp_toolkit.__gp_localid
```
Comment (darthunix) [inline src/backend/catalog/gp_toolkit.sql]: I have checked this query on Postgres planner and we get incorrect results. Port is returned from coordinator, not from segments.
```sql
adb=# select gp_execution_segment(), 'port' as param, current_setting('port') from gp_toolkit.__gp_localid;
 gp_execution_segment | param | current_setting 
----------------------+-------+-----------------
                    0 | port  | 15432
                    1 | port  | 15432
                    2 | port  | 15432
(3 rows)

adb=# explain select gp_execution_segment(), 'port' as param, current_setting('port') from gp_toolkit.__gp_localid;
                                      QUERY PLAN                                      
--------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..13500.00 rows=1000000 width=0)
   ->  Result  (cost=0.00..13500.00 rows=333334 width=0)
         ->  External Scan on __gp_localid  (cost=0.00..13500.00 rows=333334 width=0)
 Optimizer status: legacy query optimizer
(4 rows)
```

**Review by darthunix — CHANGES_REQUESTED** [review]

File: src/backend/catalog/gp_toolkit.sql (hunk 1 above)
Comment (tsdkv) [reply src/backend/catalog/gp_toolkit.sql]: Last changes return correct results

Comment (darthunix) [reply src/backend/catalog/gp_toolkit.sql]: @t1mursadykov now it works. But...

`gp_toolkit.__gp_param_local_setting()` in fact is a SQL [wrapper](https://github.com/arenadata/gpdb/blob/b71e6f99814cda1f37a825b6fd71a1c77d0860fc/src/backend/catalog/gp_toolkit.sql#L462) over your previous approach with `gp_execution_segment()`, `param` and `current_setting()`. Are you sure that we always return `param` from segments when user calls `gp_toolkit.gp_param_setting()` in a query and no session GUC can change this behaviour?

Comment (tsdkv) [reply src/backend/catalog/gp_toolkit.sql]: @darthunix In both cases the targetlist is executed on a segment.  But in first case, where port returned from coordinator, `current_setting` has `Const` expression type and `FieldSelect` in second.  This is due to the fact that this function is **stable**, optimizer calculates it while planning and use this result in runtime. see `need_tlist_eval` variable in `grouping_planner` 

In the second case, targetlist is also executed on segments, but the `__gp_param_local_setting` function is **immutable**. The optimizer makes no assumptions about the behavior of such functions. So it calculates at runtime.

With this behavior, the problem is that the `Result` node will become on top of `GatherMotion`. In this case the Result will be executed on the coordinator and then all results will be incorrect. But fortunately this is not possible. The result node is added in `set_plan_references` called in `planner.c:328`.  And the Motion node is added in `cdbparallelize` in `planner.c:348`.  This consistent execution cann't be changed by any GUC. 

However, it is important to note that a motion node can be added even before `cbdparalelize` is called (while plannig). But it seems like motion adds only over join and subplans.

Comment (darthunix) [reply src/backend/catalog/gp_toolkit.sql]: @t1mursadykov I didn't understand your explanation (sorry). We have two queries:
```sql
-- wrong
select gp_execution_segment(), 'port' as param, current_setting('port') from gp_toolkit.__gp_localid;

-- right
select (gp_toolkit.__gp_param_local_setting('port')).* from gp_toolkit.__gp_localid;
```
where `gp_toolkit.__gp_param_local_setting()` is a SQL wrapper over the wrong query. I don't see any grouping sets here (for both of the queries `need_tlist_eval` is set `true`) - can you explain in more detail its influence on the resulting plan?

> due to the fact that this function is stable, optimizer calculates it while planning
> ...
> function is immutable. The optimizer makes no assumptions about the behavior of such functions. So it calculates at runtime.

Can you show the places in the code about these statements? I was surprised to know that it works this way.

And one more question - what is about ORCA?

Comment (tsdkv) [reply src/backend/catalog/gp_toolkit.sql]: > Can you show the places in the code about these statements?

I was wrong. 

In wrong query planner evaluates the value while planning in `evaluate_function` function, which called in 
```
simplify_function (funcid=2077, result_type=25, result_typmod=-1, args=0x7fff4c879260, allow_inline=1 '\001', context=0x7fff4c87a290) at clauses.c:3266
#1  0x00005588c96463ac in eval_const_expressions_mutator (node=0x5588cc33d828, context=0x7fff4c87a290) at clauses.c:2077
...
#5  0x00005588c96483be in eval_const_expressions_mutator (node=0x5588cc33d640, context=0x7fff4c87a290) at clauses.c:2961
#6  0x00005588c964603e in eval_const_expressions (root=0x5588cc33dd68, node=0x5588cc33d640) at clauses.c:1953
#7  0x00005588c961b8fe in preprocess_expression (root=0x5588cc33dd68, expr=0x5588cc33d640, kind=1) at planner.c:880
#8  0x00005588c961b2a0 in subquery_planner (glob=0x5588cb7c9068, parse=0x5588cc39d5e8, parent_root=0x0, hasRecursion=0 '\000', tuple_fraction=0, subroot=0x7fff4c87a410, 
    config=0x5588cc33dcf0) at planner.c:656
```

As we see in the function description (`evaluate_function` ), we can evaluate functions expression if the function is immutable and has all constant inputs. Yes, thats exacly what we have with `__gp_param_local_setting`, but inside of this function we have check for set-returning-functions. For SRF we do nothing. Thats why in the right query the function will not evaluate while planning.

Comment (darthunix) [reply src/backend/catalog/gp_toolkit.sql]: @t1mursadykov thanks for the explanation about PostgreSQL planner. And what is about ORCA?

Comment (tsdkv) [reply src/backend/catalog/gp_toolkit.sql]: When wrong query is run, orca receives a function that has already been calculated ([here](https://github.com/greenplum-db/gpdb/blob/03b2ffd61e2b5f06b1e77e55e8a116f7613f0ab2/src/backend/optimizer/plan/orca.c#L128)). 

The logic is the same as for the postgres planner.
``` c
#0  evaluate_function (funcid=2077, result_type=25, result_typmod=-1, args=0x56322e912288, func_tuple=0x7f1f8a9a0720, context=0x7fff371cf840) at clauses.c:3462
#1  0x000056322d7edb89 in simplify_function (funcid=2077, result_type=25, result_typmod=-1, args=0x7fff371ce790, allow_inline=1 '\001', context=0x7fff371cf840) at clauses.c:3263
#2  0x000056322d7eb5b2 in eval_const_expressions_mutator (node=0x56322ee86a60, context=0x7fff371cf840) at clauses.c:2077
...
#8  0x000056322d7f1c50 in query_or_expression_tree_mutator (node=0x56322ee86010, mutator=0x56322d7eb2e2 <eval_const_expressions_mutator>, context=0x7fff371cf840, flags=0) at clauses.c:4886
#9  0x000056322d7eaf8e in fold_constants (glob=0x56322ebad2c8, q=0x56322ee86010, boundParams=0x0, max_size=102400) at clauses.c:1822
#10 0x000056322d7e019c in preprocess_query_optimizer (glob=0x56322ebad2c8, query=0x56322ee86010, boundParams=0x0) at transform.c:57
#11 0x000056322d7e1435 in optimize_query (parse=0x56322ebad400, boundParams=0x0) at orca.c:128
...
```
Orca minidump where we can see that orca gets the evaluated functions with value parameter (for wrong query)

```xml

             <dxl:ProjElem ColId="9" Alias="gp_execution_segment">
                    <dxl:FuncExpr FuncId="0.6022.1.0" FuncRetSet="true" TypeMdid="0.23.1.0" />
                    </dxl:ProjElem>
                    <dxl:ProjElem ColId="10" Alias="param">
                        <dxl:ConstValue TypeMdid="0.705.1.0" Value="cG9ydAA=" />
                    </dxl:ProjElem>
                    <dxl:ProjElem ColId="11" Alias="current_setting">
                        <dxl:ConstValue TypeMdid="0.25.1.0" Value="AAAACTE1NDMy" LintValue="1412543444" />
                    </dxl:ProjElem>
                </dxl:ProjList>
```

Comment (tsdkv) [reply src/backend/catalog/gp_toolkit.sql]: Also, orca doesn't support fieldselect. It is the operation of extracting one field from a tuple. 

``` sql 
create type type_B as (
   a int,
   b int 
);

set client_min_messages = 'log';
explain select (select (1,2)::type_B from (select 1) as t).* from (select 1) as t;
```
Orca returns 
```
LOG:  2021-09-24 03:23:08:019890 MSK,THD000,NOTICE,"Feature not supported: FIELDSELECT",
```

Therefore, the problem of the right query is not really relevant for orca. But in any case, even with the adding of this feature to orca, everything will work.

**Review by darthunix — APPROVED** [review]

**Pull request merged**